JSON Logging with Request ID

burnes opened the pull request:
- logback-promtail.xml config now logs in JSON format
- provide exactly the keys we are using in promtail configuration, so we don't have to parse them again
- include `requestId` in logs from provided `X-Request-ID` header. This header should be provided by ingress / proxy nginx
- Update various dependencies (similar versions also used e.g. in lovely-micronaut-graphql)

burnes committed:
Provide requestId in promtail logs

dobe commented on src/test/kotlin/ls/http/RequestIdFilterTest.kt:
```diff
@@ -0,0 +1,54 @@
+
+val elements = mutableListOf<PropagatedContextElement>()
+
+val testMutablePropagatedContext = object : MutablePropagatedContext {
```
you can use mockk for such things, however i would test this on a higher level. since you need to test the real http handling anyways.

dobe commented on CHANGES.md:
```diff
@@ -1,5 +1,19 @@
+
+### Feature
+
+- Provide requestId in promtail logs
```
as discussed, we use json in future, we should not make an incompatible master merge with a feature nobody should use

pls also do not set version numbers in feature additions, that should be part of the actual release.

burnes committed:
JSON Logging

use logstash encoder (https://github.com/logfellow/logstash-logback-encoder) and provide properties defined in current promtail config. Then we don't need to parse data anymore in promtail.
Also update some other dependencies (similar versions to lovely-micronaut-graphql)

dobe commented on src/main/kotlin/ls/http/RequestIdFilter.kt:
```diff
@@ -0,0 +1,28 @@
+ * Filter that adds the request id to the MDC propagation context.
+ * The request id is taken from the `X-Request-ID` header.
+ *
+ * The request id will be included in all logback-promtail logs.
```
i think we do not need to document the logback or promtail stuff here. i would do that in the configuration file. this method  just ensures that the request id is in the context.

dobe commented on src/main/kotlin/ls/http/RequestIdFilter.kt:
```diff
@@ -0,0 +1,28 @@
+    fun rememberRequestId(request: HttpRequest<*>, mutablePropagatedContext: MutablePropagatedContext) {
+        val trackingId = request.headers.get("X-Request-ID")
+        // trackingId or empty string
+        val context = MdcPropagationContext(mutableMapOf("requestId" to trackingId.orEmpty()))
```
do you need a mutablemap here? could probably use mapOf

dobe commented on src/main/resources/logback-promtail.xml:
```diff
@@ -1,9 +1,16 @@
 <configuration>
     <appender name="STDOUT" class="ch.qos.logback.core.ConsoleAppender">
-        <encoder>
+        <encoder class="net.logstash.logback.encoder.LogstashEncoder">
```
pls doc here ho context vars etc get into the json output

dobe commented on src/test/resources/application-test.yml:
```diff
@@ -0,0 +1,2 @@
+logger:
+  config: logback-promtail.xml
```
you can set this in the test (property annotation), no need to change the logging for all tests

dobe commented on src/test/kotlin/ls/http/LoggingTest.kt:
```diff
@@ -0,0 +1,79 @@
+import java.io.PrintStream
+
+@MicronautTest(transactional = false)
+class LoggingTest(@Client("/") httpClient: HttpClient) : FreeSpec({
```
i thing it would be better to test the request id filter sperrately and just check if it is propagated to the context.

you could still have a logging test, which in general ensures that context vars are printend, however i think we do not need to necessarily test the log format. if you think, we should renam this test to sometzhing like PromtailLogging or so, and not mess with request id

dobe commented on src/test/kotlin/ls/http/LoggingTest.kt:
```diff
@@ -0,0 +1,79 @@
+        // Create ByteArrayOutputStream to capture logs
+        val baos = ByteArrayOutputStream()
+        // Redirect System.out to the ByteArrayOutputStream
+        System.setOut(PrintStream(baos))
```
pls avoid messing around with sys streams. use a special logger.

burnes committed:
some improvements

still looking for some solutions....

burnes committed:
improved tests

burnes committed:
restrict test controllers to test class

dobe commented on settings.gradle.kts:
```diff
@@ -31,17 +32,25 @@ dependencyResolutionManagement {
+            // Logging
+            library("microutils-logging", "io.github.microutils", "kotlin-logging-jvm").version("3.0.5")
+
+            library("micronaut-jackson-databind", "io.micronaut", "micronaut-jackson-databind").version("4.4.6")
```
micronaut libs are in the mn catalog afaik and do not need to be created here

burnes replied:
true for micronaut-jackson-databind, microutils-logging is not in micronaut.

dobe reviewed (changes requested):
nice!, otherwise ack

burnes committed:
no need for jackson databind

already included with mn.jackson.module.kotlin

burnes committed:
Make the logback-promtail.xml encoder available for the runtime

so all projects using it must not manually do it

burnes committed:
do not provide version and level_value
